### PR TITLE
Make nonnull assertions and binding patterns apparent declared type locations

### LINKED ISSUE #20994: Weird uninitialized variable error for destructured members of type parameter constrained to homomorphic mapped type where member is declared optional
Found in Formik by @weswigham who patiently worked with me to find the minimal repro.

Run the following with `strictNullChecks`

```ts
interface Options {
    a?: number | object;
    b: () => void;
}

class C<T extends Options> {
    foo!: { [P in keyof T]: T[P] }

    method() {
        let { a, b } = this.foo;
        !(a && b)
        a
        // TypeScript issues an error on 'a' above:
        // Variable 'a' is used before being assigned.'
    }
}
```

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #20994

### CODE REVIEW

**Comment by ahejlsberg** [comment]
Please explain what the issue is and how this fix solves it.

**Comment by weswigham** [comment]
In the process of doing control flow analysis for `a`, we replace it with it's filtered constraints in either branch of the conditional following the destructuring pattern, then union them back together. In this way, `T["a"]` becomes `undefined | number | object`. We fire off a used before defined check because we don't see that `undefined` is in the domain of `T["a"]`. By checking it's apparent type for falsy flags, we check against the constraint instead, which yields the correct result.

**Comment by Kovensky** [comment]
This seems to fix the problem, but I don't think it's fixing the _actual_ problem, which is the "use before assignment" assertion on something that has been assigned. Regardless of what the type of `T["a"]` is, after `let { a } = anything`, `a` is definitely assigned. If `T["a"]`'s domain includes `undefined`, that just makes so that `a`'s type can include `undefined`.

**Comment by weswigham** [comment]
@Kovensky the way that we do use before assignment checks in strict null checks mode is all about the type. You can only use before assignment a type without `undefined` in it's domain, so if the type at a specific location still has `undefined` in it (which is introduced by a declaration without an assignment) but didn't have it in the declared type, then it must still be there because it hasn't been assigned yet. At least, that's the assumption we currently make to issue the error.

**Comment by mhegazy** [comment]
LGTM. @ahejlsberg any comments?

**Comment by weswigham** [comment]
@ahejlsberg This now uses the fix we talked about in person.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -13162,19 +13153,20 @@ namespace ts {
...
+                parent.kind === SyntaxKind.NonNullExpression ||
+                parent.kind === SyntaxKind.BindingElement && (<BindingElement>parent).name === node && !!(<BindingElement>parent).initializer;
         }
 
         function typeHasNullableConstraint(type: Type) {
             return type.flags & TypeFlags.TypeVariable && maybeTypeOfKind(getBaseConstraintOfType(type) || emptyObjectType, TypeFlags.Nullable);
         }
 
-        function getDeclaredOrApparentType(symbol: Symbol, node: Node) {
+        function getDeclaredOrApparentType(type: Type, node: Node) {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: We need a better name for this method.

Comment (weswigham) [reply src/compiler/checker.ts]: Renamed to `getApparentTypeForPosition`

**Comment by weswigham** [comment]
@ahejlsberg done

**Comment by ajafff** [comment]
@weswigham I think @Kovensky is right that fixing the issue mentioned above could be achieved with a smaller change.
Identifiers whose declaration is a BindingElement are always initialized. That means you could just add another condition to the initializer of `assumeInitialized` inside `checkIdentifier` (currently line 13043).

**Pull request merged**